keps/sig-node/4188-kubelet-pod-readiness-api/README.md

```diff
@@ -0,0 +1,345 @@
+| - | Kubelet does not support RBAC authorization for gRPC. (mitigation: This API is design to be accessible for all workloads running on the node without the authorization. The unix socket will be used for the connection and all exposed data will be carefully reviewed.) |
+
+### Goals
+
+The goal of this API is to expose Pod readiness information directly from the source - kubelet. This would remove the need for node-local components to request this node-local information from the Kubernetes API Server. Kubelet already has a podresources endpoint ([2403-pod-resources-allocatable-resources](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/2403-pod-resources-allocatable-resources)) which returns information about Pod’s containers and Devices. This API does not contain information about pod readiness status. Kubelet is responsible for computing pod status and stores it in a local cache. We want to create a new API that will expose pod conditions that are computed by kubelet and return the most recent data even when kube-API is not reachable.
+
+### Non-Goals
+
+### User Stories
+* Some on node system workloads want to reduce Control Plane dependency and introduce locality for pod’s readiness due to reliability and scalability issues.
+* For building custom monitoring tools that show information about pod’s readiness. When it is ready to serve the traffic.
+* Some on node system workloads interested in pods readiness want to reduce resource consumption.
+
+## Proposal
+
+We are proposing to create a new gRPC API that will return pod's readiness information.
+* The API will return data about both Static and Regular Pods.
+* The API will not return partial data. If kubelet does not know actual information about workloads then appropriate error code will be returned.
+* API should return the most recent information about pods computed by the kubelet even when those data were not reported or accepted by kube-API.
+* The API will be read-only and accessible for on-node workloads (we will use unix socket to connect) without the authorization.
+* The API will be versioned.
+
+
+### Risks and Mitigations
+
+This API is read-only, which removes a large class of risks.
+
+| Risk                                                      | Impact        | Mitigation |
+| --------------------------------------------------------- | ------------- | ---------- |
+| Too many requests to the API impacting the kubelet performances | High          | Rate limiting the API. |
+| Misuse of the API  | High |  This API is Read-only. We will expose only a small portion of the pod's information related to the pod readiness. Exposed data does not contain sensitive information that could be used in a malformed way. |
+| Unauthorized access to the API | Medium | This API is designed to be accessed by all on-node workloads. There is no need to authorize the clients to access this API. |
```

Comment:
Why not? The API server RBAC rules allow isolating namespaces for example. How will this work in a multi-tenant environments? Will all potential users of a cluster have read access to pod and container names, statuses etc of all the other customers' workloads?

Reply:
yes, this API is design to work for system pods that needs to know the state of the node (all running workloads). Adding API server RBAC rules will introduce ControlPlane dependency on this API and the reason why we propose this API is that we want to fetch the data directly form kubelet without ControlPlane dependency. 
If someone wants the isolation then Kube-API should be used and this API can be enabled/disabled with feature gate.

Reply:
>If someone wants the isolation then Kube-API should be used and this API can be enabled/disabled with feature gate.

Reply:
@ffromani good point! Since the API is read-only, I don't have the strong opinion on reading other's pod/container name and status, but we might want to consult with SIG Auth folks. I won't treat this as the blocker issue for alpha though.

Reply:
For reference, I believe both @deads2k and @liggitt provided feedback on this KEP when we met with sig-arch about this proposal on Aug 24 and Sep 7. I may be misremembering here, but I'd thought that at least Pod readiness was considered acceptable to expose to workloads running on the same Node.

Reply:
> that at least Pod readiness was considered acceptable to expose to workloads running on the same Node.

I'd like to make sure this KEP only touches the pod readiness. At least suggests more information is considered, and if so it should be explicitly specified to ensure we don't leak sensitive information